HTTP chain observer

ch1bo opened the pull request:
Extends `hydra-chain-observer` to submit observations to a `hydra-explorer` instance with the new observer API.

See https://github.com/cardano-scaling/hydra-explorer/blob/observer-backend/README.md#architecture for an architectural overview.

* :mag: Adds `--explorer` option to `hydra-chain-observer` which submits observations using HTTP. This uses the `OnChainTx` type, so there are new roundtrip tests and golden files.
* :mag: Make `hydra-chain-observer` aware of its `--version` (also via embedding git revs using nix).
* :mag: Switch on `--node-socket` or `--blockfrost-project-path` in `hydra-chain-observer` instead of subcommands. This allows us to re-use more code and the help text becomes easier to understand. See hydra-chain-observer/README.md for how to invoke the binary now.

* :mag: Extracts a `common.yaml` from API definitions so we can reference it from `hydra-explorer` API specs / schemas
* :mag: Moves observation code from `hydra-node` to `hydra-tx`, this improves dependency closure for `hydra-explorer`, especially once we move `Hydra.Chain` out of `hydra-node` (e.g. by inverting the dependency of `hydra-chain-observer`?)
  - This requires 
* Adds orphan instances to `hydra-cardano-api` as needed by the `hydra-explorer`, see https://github.com/cardano-scaling/hydra-explorer/pull/18

TODO:
- [x] Client-side serialization tests against observer openapi

---

* [x] CHANGELOG updated
* [x] Documentation updated (README, no user guide)
* [x] Haddocks updated
* [x] No new TODOs

ch1bo committed:
Move HeadSeed and HeadId conversion to hydra-tx

ch1bo committed:
Move not used functions into Hydra.Chain.Direct.State

ch1bo committed:
Move tx observation to hydra-tx

This allows the observation types to be easier re-used.

ch1bo committed:
Extract a common json schema to be referenced

ch1bo committed:
Draft desired behavior to report observations

ch1bo committed:
Use options instead of subcommands in hydra-chain-observer

This makes documentation and implementation for common options like
"--start-chain-from" and (new) "--explorer" less redundant.

ch1bo committed:
Add --explorer to hydra-chain-observer

ch1bo committed:
Perform POST requests on observation

ch1bo committed:
Change ChainObservation to match observer API json schema

ch1bo committed:
Add FromJSON OnChainTx

This type is used by hydra-chain-observer and hydra-explorer
respectively on the observations API

ch1bo committed:
Add HydraVersion to common.yaml

ch1bo committed:
Mark necessary changes as FIXME and optional as TODO

ch1bo committed:
Add ToJSON/FromJSON orphan instances for NetworkMagic

Cardano-api has instances for Network, but not NetworkMagic.

ch1bo committed:
Add Arbitrary NetworkId instance to hydra-cardano-api

ch1bo committed:
Add roundtrip golden tests for OnChainTx

This data type is essential for the hydra-chain-observer ->
hydra-explorer API

ch1bo committed:
Submit version and network to hydra-explorer

The hydra-chain-observer now uses the hydraNodeVersion, which is either
patched into it or uses the git revision from the repository, and
reports this along with a network id to the explorer.

ch1bo committed:
Update changelog

ffakenz commented on hydra-chain-observer/test/Hydra/ChainObserverSpec.hs:
```diff
@@ -3,19 +3,30 @@ module Hydra.ChainObserverSpec where
+    -- NOTE: Detect regressions in interface to hydra-explorer
+    roundtripAndGoldenADTSpecsWithSettings
+      defaultSettings{sampleSize = 1}
+      $ Proxy @(OnChainTx Tx)
```
maybe we should be doing this test using the ChainObservation type.

ch1bo committed:
Update changelog and hydra-chain-observer version

Using patch version component as this is not really changing the main
hydra features, but only the reporting into the hydra-explorer.

ch1bo committed:
Add weeder roots for orphan instances

ch1bo committed:
Add json-schemas/common.yaml to cabal data-files

Out of source tree builds/tests are only provided the files listed.
